### PR TITLE
[Refactor]: consistent update of bufferline offset with nvim-tree

### PR DESCRIPTION
**Pull request by ahmedkhalf**
# Description

This adds buffer local autocmds to setup the bufferline offset which is
important since currently the bufferline doesn't get updated when someone does
`NvimTreeOpen` and co. Also important when someone does `:q`.

This also refactor the changedir function.

## How Has This Been Tested?

I opened and closed nvimtree

### CODE REVIEW

**Comment by ahmedkhalf** [comment]
Before the merge, please discuss on discord.

**Review by Tastyep — CHANGES_REQUESTED** [review]
There is a case where it doesn't work
Open the tree, find a new file so that it opens a new tab
close and open the tree until you see the problem (tabs don't realign to the left)

**Comment by ahmedkhalf** [comment]
https://user-images.githubusercontent.com/36672196/130043551-bf1d967a-824d-44b6-825a-25038e933527.mp4

Cannot reproduce

File: lua/core/nvimtree.lua
Comment (kylo252) [inline lua/core/nvimtree.lua:88]: Is the use-case worth adding 3 additional auto-commands trying to fix a limitation in nvim-tree? are there no performance penalties for using this [vimL] ? I don't think anyone uses `NvimTreeOpen`. And I'm not sure what the problem is with `:q`.

I feel like all of this can be just as easily added in `config.lua`.

Comment (ahmedkhalf) [reply lua/core/nvimtree.lua:88]: And honestly for peace of mind, I don't mind that

Comment (kylo252) [reply lua/core/nvimtree.lua:88]: By the way, I still don't fully get the problem with `:q`.

So nvim-tree doesn't handle the bufferline, that's up to barbar which lists the following as a way to handle this: https://github.com/romgrk/barbar.nvim#integration-with-filetree-plugins

> And honestly for peace of mind, I don't mind that

How is it a peace of mind to add a hook that gets evaluated every time you open nvim-tree, but the same hook is not actually maintained by nvim-tree?

Comment (ahmedkhalf) [reply lua/core/nvimtree.lua:88]: The problem is not `:q` nor is it `:NvimTreeOpen`, the problem is a combination of everything that can open or close nvim-tree without calling `M.toggle_tree`.

Comment (kylo252) [reply lua/core/nvimtree.lua:88]: > The problem is not `:q` nor is it `:NvimTreeOpen`, the problem is a combination of everything that can open or close nvim-tree without calling `M.toggle_tree`.

I still don't know what the _problem that we're trying to fix_ is. My thought is that you don't have to use that binding, and just use the default ones instead.  

```lua
-- just add this to you config.lua
lvim.builtin.which_key.mappings["e"] = { "<cmd>NvimTreeToggle<CR>", "Explorer" }
```

> But if you want this to be inside the nvim-tree codebase, then sure, I could create an issue there for them to create a callback.

If it's an **nvim-tree** problem that's not related to **barbar** specifically, then yeah I would say it's a better fit there instead.

Comment (ahmedkhalf) [reply lua/core/nvimtree.lua:88]: > I still don't know what the _problem that we're trying to fix_ is. 

I was pretty clear, but let me show you instead.

https://user-images.githubusercontent.com/36672196/130056420-69a5a43d-b238-4778-9cd4-abc8547bc48e.mp4

**Comment by Tastyep** [comment]
> simplescreenrecorder-2021-08-19_13.17.47.mp4
> 
> Cannot reproduce

Please redo the same test, but this time open a file before opening the tree so that you get two tabs after opening the second file

**Comment by ahmedkhalf** [comment]
> > simplescreenrecorder-2021-08-19_13.17.47.mp4
> > Cannot reproduce
> 
> Please redo the same test, but this time open a file before opening the tree so that you get two tabs after opening the second file

Still cannot reproduce, I tried opening the first file from telescope and with lvim [file].

**Comment by Tastyep** [comment]
> > > simplescreenrecorder-2021-08-19_13.17.47.mp4
> > > Cannot reproduce
> > 
> > 
> > Please redo the same test, but this time open a file before opening the tree so that you get two tabs after opening the second file
> 
> Still cannot reproduce, I tried opening the first file from telescope and with `lvim [file]`.

Try the following steps:
- Open `init.lua`
- `<space>e` <- you are now in the tree
- `<space>f` whichkey 
- 3 x `<space>e`

File: lua/core/nvimtree.lua
Comment (kylo252) [reply lua/core/nvimtree.lua:88]: That's related to Dashboard and not nvim-tree. Try to use the "glepnir/dashboard-nvim" instead of Chris' fork and you'll see that the problem is no longer there. I'm not sure if the fork is still necessary btw.

Xzz4gsFdAh

What about the `:q` issue?

Comment (ahmedkhalf) [reply lua/core/nvimtree.lua:88]: Dude... wtf. I don't care if the tabs show or hide on dhasboard. My point is the same offset bug is visible on a normal file...

**Comment by ahmedkhalf** [comment]
Finally able to reproduce, turns out it happens when you wait for whichkey popup. Will see what I can do about it.

**Comment by ahmedkhalf** [comment]
I just realized there is a bug with the icons, please don't merge until I fix it.

**Comment by ahmedkhalf** [comment]
Awaiting final review and merge

File: lua/core/nvimtree.lua
Comment (Tastyep) [inline lua/core/nvimtree.lua:101]: `and` instead of nesting `if`s

**Review by Tastyep — CHANGES_REQUESTED** [review]

**Review by Tastyep — APPROVED** [review]

File: lua/core/nvimtree.lua
Comment (Tastyep) [inline lua/core/nvimtree.lua:98]: Can you remove this for a newline :)

File: lua/core/nvimtree.lua
Comment (Tastyep) [inline lua/core/nvimtree.lua:106]: here too

**Review by Tastyep — CHANGES_REQUESTED** [review]

File: lua/core/nvimtree.lua
Comment (ahmedkhalf) [reply lua/core/nvimtree.lua:98]: you mean this `--` ?

Comment (Tastyep) [reply lua/core/nvimtree.lua:98]: Yes, please :)

Comment (Tastyep) [reply lua/core/nvimtree.lua:98]: Just make it a standard blank line

**Review by Tastyep — APPROVED** [review]

**Pull request merged**
